MNG-6771 Fix license issues on binary distribution

eolivelli opened the pull request:
Draft PR for https://issues.apache.org/jira/browse/MNG-6771

- Start fixing the template for LICENSE file
- Have consistent naming for LICENSE file and final jars by removing 'classifier' from the name of jars
- Introduce a script to check for license, from Apache BookKeeper project, original author ivank@apache.org

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [ ] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [ ] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

eolivelli commented on apache-maven/src/main/appended-resources/META-INF/LICENSE.vm:
```diff
@@ -22,21 +22,18 @@ Apache Maven includes a number of components and libraries with separate
 ##
-#set ( $apacheLicenseNames = [ "Apache License, Version 2.0", "The Apache Software License, Version 2.0",
-    "ASLv2", "Apache Public License 2.0", "Apache 2.0" ] )
+#set ( $apacheMavenGroupIds = [ "org.apache.maven", "org.apache.maven.wagon", "org.apache.maven.resolver",
```
Skip our stuff, and all ASLv2 stuff from anyone
see http://www.apache.org/dev/licensing-howto.html#alv2-dep

eolivelli commented on apache-maven/src/main/appended-resources/META-INF/LICENSE.vm:
```diff
@@ -22,21 +22,18 @@ Apache Maven includes a number of components and libraries with separate
 #*    *##set ( $spdx = false )
 #*    *##set ( $includeLicense = true )
 #*    *###
-#*    *##if ( $project.artifact.artifactId == "jcl-over-slf4j" )
```
this is to be discussed, if we should really report a different license other the one that is reported on Maven Central

michael-o replied:
This must be raised upstream.

hboutemy replied:
everything already done: license in pom.xml was not consistent with license in scm
see https://issues.apache.org/jira/browse/MNG-6779 and the associated PR to slf4j that has already been merged

michael-o replied:
This means we have to wait for 1.7.29, don't we?

hboutemy replied:
we need to wait for a fixed release to remove the hack that overrides the license: the hack is not so big, we can live with it for now
(notice the PR https://github.com/qos-ch/slf4j/pull/226 was merged in 2.0.0-alpha branch only)

michael-o replied:
That's bad and needs to be backported to 1.7.x line.

eolivelli replied:
The real problem is that we can change here the output in LICENSE but we can't change the license file we download to 'lib'.
So they won't be consistent

eolivelli commented on apache-maven/src/main/assembly/component.xml:
```diff
@@ -29,6 +29,7 @@ under the License.
     <dependencySet>
       <useProjectArtifact>false</useProjectArtifact>
       <outputDirectory>lib</outputDirectory>
+      <outputFileNameMapping>${artifact.artifactId}-${artifact.version}.${artifact.extension}</outputFileNameMapping>
```
This is to have a cleaner and consistent naming of jars in LICENSE file and in 'lib' directory.
We should add 'classifier' in both or drop it in both, actually it seems simpler to do drop

hboutemy replied:
I don't like this: why do you find it useful?

eolivelli replied:
In order to make the names of the jar files look the same as we write them in the LICENSE file.

I can try to change the LICENSE file to include the qualifier and revert this line

eolivelli replied:
reverted

eolivelli commented on dev/check-binary-license:
```diff
@@ -0,0 +1,117 @@
+set -e
+
+TARBALL="$1"
+if [ -z $TARBALL ]; then
```
This is just the adaptation of 
https://github.com/apache/bookkeeper/blob/master/dev/check-binary-license

That in turn verifies for
http://bookkeeper.apache.org/community/licensing

We will adapt it to whatever rules we decide for Maven

we could also convert to a Groovy/Python script,
honestly I would love to see these kind of checks in the "Apache RAT" project

eolivelli commented on dev/check-binary-license:
```diff
@@ -0,0 +1,117 @@
+if [ $EXIT != 0 ]; then
+    echo
+    echo "It looks like there are issues with the LICENSE/NOTICE (error $EXIT)".
+    echo See http://bookkeeper.apache.org/community/licensing for details on how to fix.
```
this is to be changed

michael-o commented on dev/check-binary-license:
```diff
@@ -0,0 +1,117 @@
+
+TAR='tar'
+unamestr=`uname`
+if [[ "$unamestr" == 'Linux' ]]; then
```
That's bad code. You should rather test for GNU tar instead of Linux.

eolivelli replied:
Ok, I am not a bash expert, thanks for your suggestion. I will try to come with a better check

eolivelli replied:
I am dropping this file

eolivelli commented:
@hboutemy I have fixed the download of ASL license files.
I have also added a tweak to bundle jsoup license: add an explicit dependency and then exlucde the jar file from the "lib" directory.

I can't find the point in code where we are generating the NOTICE file for the Binary distribution. 
It is not a raw source file copied from somewhere.

I need to be able to manually write this file.
This is the last step in this work

eolivelli commented:
@hboutemy  when I am referring to '.license' files I am talking about this line:
https://github.com/apache/maven/blob/master/apache-maven/src/main/appended-resources/META-INF/LICENSE.vm#L54

that makes that the license files are downloaded to lib/xxxx.license

hboutemy commented:
> I have also added a tweak to bundle jsoup license: add an explicit dependency and then exlucde the jar file from the "lib" directory.

nice tweak :)

> I can't find the point in code where we are generating the NOTICE file for the Binary distribution.

it comes from maven-remote-resources-plugin processing org.apache:apache-jar-resource-bundle:1.4, cf https://github.com/apache/maven-resources/tree/trunk/apache-jar-resource-bundle/src/main/resources/META-INF

eolivelli commented:
Thank you @hboutemy  for moving forward.
The last step is to create the NOTICE file.
Are you working on it ?
I have time to do the first step this evening. I will go thru the list of depedencies and collect manually all of the texts to be reported in NOTICE file

eolivelli commented:
@hboutemy I have written on JIRA the contents I think we need to write in NOTICE
is there any way to append contents to it comes from maven-remote-resources-plugin processing org.apache:apache-jar-resource-bundle:1.4, cf https://github.com/apache/maven-resources/tree/trunk/apache-jar-resource-bundle/src/main/resources/META-INF 

without releasing a new version of maven-resources ?

hboutemy commented:
@eolivelli if you copy MANIFEST.vm to NOTICE.vm, you'll see that you can do the same Velocity append in NOTICE than what is currently done for MANIFEST

hboutemy commented:
ok for me

eolivelli commented:
Thank you @hboutemy 
I will merge as soon as CI is blue https://builds.apache.org/job/maven-box/job/maven/job/MNG-6771/15/

eolivelli commented:
committed as f8bd6d577e43ba0c81f5be240d5282caf780f18d

thank you so much @hboutemy  !

jira-importer commented:
Resolve #8496
